scheduler/README.md

```diff
@@ -0,0 +1,192 @@
+                                                         └───────────┘     └──────────────┘     └─────────────┘                           └──────────┘
+```
+
+## Cluster reconciliation
+
+The first step for the service and bach job scheduler is called
+"reconcilliation," and its logic lies in `scheduler/reconcile.go` file. The
+`allocReconciler` object has one public method: `Compute`, which takes no
+arguments and returns `reconcileResults` object. This results object tells the
+scheduler about desired deployment to be updated or created, which allocations
+to place, which should be updated destructively or in-place, which should be
+stopped, and which are disconnected or are reconnecting. The reconciler works
+in terms of "buckets," that is, it processes allocations by putting them into
+different sets, and that's how its whole logic is implemented.
+
+The following vocabulary is used by the reconciler:
+
+- "tainted node:" a node is considered "tainted" if allocations must be migrated
+off of it. These are nodes that are draining or have been drained, but also
+nodes that are disconnected and should be used to calculated reconnect timeout.
+The cluster reconciler commonly also refers to "untainted" allocations, i.e.,
+those that do not require migration and are not on disconnected or reconnecting
+nodes.
+
+- "paused deployment:" a deployment is paused when it has an explicit `paused`
+status, but also when it's pending or initializing.
+
+- "migrating allocations:" allocations that are on nodes that are draining.
+
+- "lost allocations:" allocations that have expired or exist on lost nodes.
+
+- "disconnecting allocations:" allocations that are on disconnected nodes
+which haven't been considered "lost" yet, that is, they are in their reconnect
+timeout.
+
+- "reconnecting allocations:" allocations on nodes that have reconnected.
+
+- "ignore allocations:" allocations which are in a noop state, the reconciler
+will not be touching these.
+
+- "expiring allocations:" allocaions which are not possible to reschedule, due
+to lost configurations of their disconnected clients.
+
+The following diagram illustrates the logic flow of the cluster reconciler:
+
+```
+          ┌─────────┐
+          │Compute()│
+          └─────────┘
+               │
+               ▼
+        ┌────────────┐
+        │create a new│          allocMatrix is created from existing
+        │ allocation │          allocations for a job, and is a map of
+        │  matrix m  │          task groups to allocation sets.
+        └────────────┘
+               │
+               ▼                deployments are unneeded in 3 cases:
+ ┌───────────────────────────┐  1. when the are already successful
+ │cancelUnneededDeployments()│  2. when they are active but reference an older job
+ └───────────────────────────┘  3. when the job is marked as stopped, but the
```

Comment:
This line isn't finished:

> but the

"but the deployment was not yet stopped"?

Reply:
oof yeah, some monodraw snafu.